hxckr-org/hxckr-test-runners
# feat: implementing stage based test cases

This PR introduces changes to test runner system, focusing on optimizing test file management and execution. The main changes include implementing a singleton TestRepoManager for efficient test repository handling, improving Docker test execution, and streamlining the test output processing.

## Key Changes

### 1. Test Repository Management
- Implemented `TestRepoManager` as a singleton class to maintain a single, persistent copy of the test repository
- Replaced repeated clone operations with an efficient git pull-based update mechanism
- Added periodic updates (every 5 minutes) to ensure test files are current
- Improved error handling and logging for repository operations

### 2. Test File Loading
- Moved test file loading logic to a dedicated system
- Added support for language-specific test extensions
- Improved error handling for missing test files
- Added structured logging for better debugging

### 3. Docker Integration
- Enhanced Docker test execution to properly handle test failures
- Improved output parsing and error handling
- Added cleanup procedures for Docker resources

### 4. Test Output Processing
- Added functions to clean and parse test output
- Implemented better success/failure detection based on test results
- Removed duplicate output in logs

## Configuration Requirements
Add the following environment variables:
```bash
TEST_REPO_URL=https://github.com/your-org/test-repo.git
TEST_REPO_BRANCH=main
TEST_REPO_PATH=tests
```

## Directory Structure
> remember to create your test repository.
Tests should be organized in the test repository as follows(tests usually starts at stage 2):
```
tests/
  challenge-id-1/
    typescript/
      stage1.test.ts
      stage2.test.ts
    rust/
      stage1.test.rs
  challenge-id-2/
    ...
```

## Testing
- Tested with multiple challenge stages
- Verified Docker test execution
- Confirmed proper cleanup of resources
- Validated test output parsing

## Breaking Changes
None. This implementation maintains backward compatibility while improving the underlying infrastructure.

## Future Improvements
1. Add support for more programming languages
2. Implement caching for frequently accessed test files
3. Add metrics collection for performance monitoring
4. Enhance test result reporting with more detailed statistics

## How to Test
1. Set up required environment variables
2. Run test suite with different challenge stages
3. Verify test repository updates
4. Check Docker execution and cleanup
5. Validate test results output
